File: algorithms/cpp/630.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

class Solution {
private:
    static bool compare(const vector<int> &a, const vector<int> &b) {
        return a[1] < b[1];
    }
public:
    int scheduleCourse(vector<vector<int>> &courses) {
        sort(courses.begin(), courses.end(), compare);
        int s = 0;
        priority_queue<int> q;
        for (auto course: courses) {
            if (s + course[0] <= course[1]) {
                s += course[0];
                q.push(course[0]);
            }
            else if (course[0] < q.top() && s - q.top() + course[0] <= course[1]) {
                s = s - q.top() + course[0];
                q.pop();
                q.push(course[0]);
            }
        }
        return q.size();
    }
};

int main() {
    return 0;
}

File: algorithms/python3/630.py
import heapq

class Solution:
    def scheduleCourse(self, courses):
        """
        :type courses: List[List[int]]
        :rtype: int
        """
        courses.sort(key=lambda x: x[1])
        s = 0
        pq = []
        for c in courses:
            if s + c[0] <= c[1]:
                heapq.heappush(pq, -c[0])
                s += c[0]
            elif len(pq):
                t = abs(pq[0])
                if t > c[0] and s - t + c[0] <= c[1]:
                    s = s - t + c[0]
                    heapq.heapreplace(pq, -c[0])
        return len(pq)
